core/cache/charmconfigwatcher.go

```diff
@@ -0,0 +1,279 @@
+	if !w.waitInitOrDying() {
+		return
+	}
+
+	name, okUnit := msg.(string)
+	if !okUnit {
+		logger.Errorf("programming error; branch deleted message was not of expected type, string")
+		return
+	}
+
+	if w.branchName != name {
+		return
+	}
+
+	// The branch we are tracking was deleted.
+	// One of the following scenarios will ensue:
+	// 1) The branch was aborted and this is the only message we will receive.
+	//    Clearing the branch info and checking whether to notify is fine.
+	//    We will send at most one notification.
+	// 2) The branch was committed and this is the first related message.
+	//    There will be a subsequent message for the master settings change.
+	//    If the branch changes mutate master config, there will be 2
+	//    notifications sent.
+	// 3) The branch was committed and this is the second related message,
+	//    coming after the master settings change message.
+	//    The first message should have resulted in no notification,
+	//    because the new master is the same as master + branch deltas.
+	//    When the branch info is removed config remains unchanged,
+	//	  so there is no notification for the second message wither.
+
+	// TODO (manadart 2019-06-18): A fix for case (2) above would be to change
+	// the cache worker so that a completed branch is not sent as a deletion,
+	// rather as a normal change. We could then detect committed vs aborted and
+	// ensure that at most one notification is sent.
+	// We would check for notification on aborted, but ignore commits,
+	// relying on the master settings change message to determine whether to
+	// notify. After processing the update, we would then evict the branch.
+
+	w.branchName = ""
+	w.branchDeltas = nil
+	w.checkConfig()
+}
+
+// checkConfig generates a new hash based on current effective configuration.
+// If the hash has changed, a notification is sent.
+func (w *CharmConfigWatcher) checkConfig() {
+	changed, err := w.setConfigHash()
+	if err != nil {
+		logger.Errorf("generating hash for charm config: %s", errors.ErrorStack(err))
+		return
+	}
+	if changed {
+		w.notify()
+	}
+}
+
+// checkConfig applies any known branch deltas to the master charm config,
+// Then compares a hash of the result with the last known config hash.
+// The boolean return indicates whether the has has changed.
+func (w *CharmConfigWatcher) setConfigHash() (bool, error) {
+	cfg := copyDataMap(w.masterSettings)
+	for _, delta := range w.branchDeltas {
+		switch {
+		case delta.IsAddition(), delta.IsModification():
+			cfg[delta.Key] = delta.NewValue
+		case delta.IsDeletion():
+			delete(cfg, delta.Key)
+		}
+	}
+
+	newHash, err := hash(cfg)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if w.configHash == newHash {
+		return false, nil
+	}
+
+	w.configHash = newHash
+	return true, nil
+}
+
+// waitInitOrDying returns true when the watcher is fully initialised,
+// or false if it is dying.
+func (w *CharmConfigWatcher) waitInitOrDying() bool {
+	select {
```

Comment:
I am not sure whether the runtime evaluates the cases in order of appearance or randomizes the order. What would happen if **both** channels can be read from? Could this cause us to potentially consume events even though the watcher is shutting down?

Reply:
I don't think this is a big concern. 

There is that race where you could process one more final message right at a call to `Kill`, but it will never be more than that, as tear-down unsubs all the listeners.